plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java

```diff
@@ -0,0 +1,150 @@
+	private JMXConnector connector;
+	private MBeanServerConnection connection;
+	private Root root;
+	private boolean isLoading;
+	private boolean isConnected;
+	private IServer server;
+	private OpenshiftConnectionProvider provider;
+	private JMXConnectorProvider factory;
+
+	public OpenshiftConnectionWrapper(OpenshiftConnectionProvider provider, IServer server,
+			JMXConnectorProvider factory) throws MalformedURLException {
+		this.provider = provider;
+		this.server = server;
+		this.factory = factory;
+		this.isConnected = false;
+		this.isLoading = false;
+	}
+
+	public IConnectionProvider getProvider() {
+		return provider;
+	}
+
+	public String getName() {
+		return "Connection for " + server.getId();
+	}
+
+	public boolean canControl() {
+		return true;
+	}
+
+	public synchronized void connect() throws IOException {
+		try {
+			IDeploymentConfig dc = OpenShiftServerUtils.getDeploymentConfig(server);
+			DebuggingContext debugContext = OpenShiftDebugUtils.get().getDebuggingContext(dc);
+			HashMap<String, Object> env = new HashMap<>();
+			env.put(JMXConnector.CREDENTIALS,
+					new String[] { debugContext.getAdminUsername(), debugContext.getAdminPassword() });
+			IPod firstPod = OpenShiftDebugUtils.get().getFirstPod(dc);
+			if (firstPod != null) {
+				int port = getLocalAdminPort(firstPod);
+				String url = "service:jmx:remoting-jmx://localhost:" + port;
+				// try to connect
+				connector = factory.newJMXConnector(new JMXServiceURL(url), env);
+				connector.connect();
+				connection = connector.getMBeanServerConnection();
+				isConnected = true;
+				provider.fireChanged(this);
+			}
+		} catch (CoreException e) {
+			throw new IOException(e);
+		}
+	}
+
+	public synchronized void disconnect() throws IOException {
+		// close
+		root = null;
+		isConnected = false;
+		try {
+			if (connector != null) {
+				connector.close();
+			}
+		} finally {
+			provider.fireChanged(this);
+		}
+		connector = null;
+		connection = null;
+	}
+
+	public boolean isConnected() {
+		return isConnected;
+	}
+
+	public Root getRoot() {
+		return root;
+	}
+
+	public void loadRoot(IProgressMonitor monitor) throws CoreException {
+		if (isConnected && root == null && !isLoading) {
+			try {
+				isLoading = true;
+				root = NodeUtils.createObjectNameTree(this, monitor);
+			} finally {
+				isLoading = false;
+			}
+		}
+	}
+
+	public void run(IJMXRunnable runnable) throws JMXException {
+		try {
+			runnable.run(connection);
+		} catch (Exception e) {
+			IStatus s = new Status(IStatus.ERROR, JMXActivator.PLUGIN_ID,
+					JMXCoreMessages.DefaultConnection_ErrorRunningJMXCode, e);
+			throw new JMXException(s);
+		}
+	}
+
+	public void run(IJMXRunnable runnable, HashMap<String, String> prefs) throws JMXException {
+		run(runnable);
+	}
+
+	private int getLocalAdminPort(IPod pod) {
+		return PortForwardingUtils.getForwardablePorts(pod).stream().filter(p -> p.getRemotePort() == 9999)
```

Comment:
can we pelase extract this to a constant (in OpenShiftDebugUtils?) and (re)use in
* here
* OpenShiftLaunchController
* OpenShiftDebugUtils